crates/ruff_db/src/file_system/vendored.rs

```diff
@@ -0,0 +1,353 @@
+    }
+
+    fn lookup_path(&mut self, path: &NormalizedVendoredPath) -> Result<ZipFile> {
+        Ok(self.0.by_name(path.as_str())?)
+    }
+}
+
+/// A path that has been normalized via the `normalize_vendored_path` function.
+///
+/// Trailing slashes are normalized away by `camino::Utf8PathBuf`s,
+/// but trailing slashes are crucial for distinguishing between
+/// files and directories inside zip archives.
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct NormalizedVendoredPath(String);
+
+impl NormalizedVendoredPath {
+    fn with_trailing_slash(mut self) -> Self {
+        self.0.push('/');
+        self
+    }
+
+    fn has_trailing_slash(&self) -> bool {
+        self.0.ends_with('/')
+    }
+
+    fn as_str(&self) -> &str {
+        self.0.as_str()
+    }
+}
+
+/// Normalizes the path by removing `.` and `..` components.
+///
+/// ## Panics:
+/// If a path with an unsupported component for vendored paths is passed.
+/// Unsupported components are path prefixes,
+/// and path root directories appearing anywhere except at the start of the path.
+fn normalize_vendored_path(path: &FileSystemPath) -> NormalizedVendoredPath {
+    debug_assert!(
+        !path.as_std_path().is_absolute(),
+        "Attempted to lookup an absolute filesystem path relative to a vendored zip archive"
+    );
+    let mut normalized_parts = camino::Utf8PathBuf::new();
+
+    // Allow the `RootDir` component, but only if it is at the very start of the string.
+    let mut components = path.components().peekable();
+    if let Some(camino::Utf8Component::RootDir) = components.peek() {
+        components.next();
+    }
+
+    for component in components {
+        match component {
+            camino::Utf8Component::Normal(part) => normalized_parts.push(part),
+            camino::Utf8Component::CurDir => continue,
+            camino::Utf8Component::ParentDir => {
+                normalized_parts.pop();
+            }
+            unsupported => panic!("Unsupported component in a vendored path: {unsupported}"),
+        }
+    }
+
+    let mut normalized = normalized_parts.into_iter().join("/");
+    // Restore the trailing slash if there was one (camino normalizes these away),
+    // as this is how zip archives distinguish between files and directories.
+    if path.as_str().ends_with('/') {
+        normalized.push('/');
+    }
+
+    NormalizedVendoredPath(normalized)
+}
+
+#[cfg(test)]
+mod tests {
+    use std::fs;
+    use std::io::Write;
+
+    use once_cell::sync::Lazy;
+    use tempfile::TempDir;
+    use zip::{write::FileOptions, CompressionMethod, ZipWriter};
+
+    use super::*;
+
+    const FUNCTOOLS_CONTENTS: &str = "def update_wrapper(): ...";
+    const ASYNCIO_TASKS_CONTENTS: &str = "class Task: ...";
+
+    static MOCK_ZIP_ARCHIVE: Lazy<Arc<[u8]>> = Lazy::new(|| {
+        let td = TempDir::new().unwrap();
+        let typeshed_path = td.path().join("typeshed");
+        let typeshed = fs::File::create(&typeshed_path).unwrap();
```

Comment:
I don't think we need a file here. You can use a `Vec<u8>` as a target and just write into that. 

But what could be simpler (and closer to the actual code) is to create a zip file manually  and commit it. You can then use `include_bytes!` to include the bytes. 

This also avoids creating a zip file on every test run.

Reply:
> But what could be simpler (and closer to the actual code) is to create a zip file manually and commit it. You can then use `include_bytes!` to include the bytes.
> 
> This also avoids creating a zip file on every test run.

I considered this. But I thought it made for a much more readable test if you could see exactly which files are being put in the mocked zip as part of the test setup. Otherwise it feels unclear what exactly we're testing.

Reply:
That makes sense. But we should replace `tempfile` with a `&mut Vec<u8>`

Reply:
Yeah, I'll make that change for sure!